csharp/src/Ice/InputStream-Class.cs

```diff
@@ -311,90 +359,217 @@ private AnyClass ReadInstance(int index)
+
+                    // We cannot read the indirection table at this point as it may reference the new instance that is
+                    // not created yet.
+
+                    IClassFactory? factory = null;
+                    if (typeId != null)
+                    {
+                        Debug.Assert(compactId == null);
+                        factory = Communicator.FindClassFactory(typeId);
+                    }
+                    else if (compactId is int compactIdValue)
+                    {
+                        factory = Communicator.FindClassFactory(compactIdValue);
+                    }
+
+                    if (factory != null)
+                    {
+                        instance = factory.Create();
+                    }
+                    else if (SkipSlice(typeId, compactId)) // Slice off what we don't understand.
+                    {
+                        instance = new UnknownSlicedClass();
+                        // Don't read the indirection table as it's the last entry in DeferredIndirectionTableList11.
+                        readIndirectionTable = false;
+                    }
+                }
+                while (instance == null);
 
-                // We cannot read the indirection table at this point as it may reference the new instance that is
-                // not created yet.
+                // Add the instance to the map/list of instances. This must be done before reading the instances (for
+                // circular references).
+                _instanceMap.Add(instance);
 
-                IClassFactory? factory = null;
-                if (typeIdOpt is string typeId)
+                // Read all the deferred indirection tables now that the instance is inserted in _instanceMap.
+                if (_current.DeferredIndirectionTableList11?.Count > 0)
                 {
-                    Debug.Assert(_current.SliceCompactId == null);
-                    factory = Communicator.FindClassFactory(typeId);
+                    int savedPos = _pos;
+
+                    Debug.Assert(_current.Slices?.Count == _current.DeferredIndirectionTableList11.Count);
+                    for (int i = 0; i < _current.DeferredIndirectionTableList11.Count; ++i)
+                    {
+                        int pos = _current.DeferredIndirectionTableList11[i];
+                        if (pos > 0)
+                        {
+                            _pos = pos;
+                            _current.Slices[i].Instances = Array.AsReadOnly(ReadIndirectionTable());
+                        }
+                        // else remains empty
+                    }
+
+                    _pos = savedPos;
                 }
-                else if (_current.SliceCompactId is int compactId)
+
+                if (readIndirectionTable)
                 {
-                    factory = Communicator.FindClassFactory(compactId);
+                    ReadIndirectionTableIntoCurrent();
```

Comment:
Couldn't this call be moved to `InputStream.IceReadStartFirstSlice`?

Reply:
It's possible. See comment below.